dashboard/api.py

```diff
@@ -367,3 +376,107 @@ def format_courserun_for_dashboard(course_run, status_for_user, certificate=None
         # TODO: here goes the logic to pull the current grade  # pylint: disable=fixme
         pass
     return formatted_run
+
+
+def get_student_enrollments(user, edx_client, course_ids):
+    """
+    Return cached enrollment data or fetch enrollment data first if necessary. Only enrollments with a course key in
+    course_ids are updated on the database, others are left alone.
+
+    Args:
+        user (django.contrib.auth.models.User): A user
+        edx_client (EdxApi): EdX client to retrieve enrollments
+        course_ids (list of str): List of course keys
+    Returns:
+        Enrollments: an Enrollments object from edx_api. This may contain more enrollments than correspond to
+            course_ids if more exist from edX, or it may contain fewer enrollments if they
+            don't exist for the course id.
+    """
+    # Data in database is refreshed after 5 minutes
+    now = datetime.datetime.now(tz=pytz.utc)
+    five_minutes_ago = now - datetime.timedelta(0, 0, 0, 0, 5)
+    enrollments_list = list(models.Enrollment.objects.filter(
+        user=user,
+        last_request__gt=five_minutes_ago,
+        course_run__edx_course_key__in=course_ids,
+    ))
+    if len(enrollments_list) == len(course_ids):
+        return Enrollments([
+            enrollment.data for enrollment in enrollments_list
+        ])
+
+    # Data is not available in database or it's expired. Fetch new data.
+    enrollments = edx_client.enrollments.get_student_enrollments()
+
+    # Make sure all enrollments are updated atomically. It's still possible that this function executes twice and
+    # we fetch the data from edX twice, but the data shouldn't be half modified at any point.
+    with transaction.atomic():
+        # Remove course ids which don't have corresponding course runs.
+        course_ids = CourseRun.objects.filter(edx_course_key__in=course_ids).values_list("edx_course_key", flat=True)
+        course_ids = set(course_ids)
+        # Delete all enrollments corresponding to course_ids
+        models.Enrollment.objects.filter(user=user, course_run__edx_course_key__in=course_ids).delete()
+
+        for enrollment in enrollments.enrollments.values():
+            if enrollment.course_id in course_ids:
+                models.Enrollment.objects.create(
+                    user=user,
+                    course_run=CourseRun.objects.get(edx_course_key=enrollment.course_id),
+                    data=enrollment.json,
+                    last_request=now,
+                )
```

Comment:
We would need to keep track of enrollments which are updated and then delete the other ones, though that solution probably performs slightly better